File: sdk/symbol_dart_sdk/bin/Network.dart
import './NetworkTimestamp.dart';
import './CryptoTypes.dart';
import 'package:hash/hash.dart';

import 'dart:typed_data';

class Network {
  String name;
  int identifier;
  NetworkTimestampDatetimeConverter datetimeConverter;
  Function addressHasher;
  Function createAddress;
  Type addressType;
  Type networkTimestampClass;

  Network(
    this.name, 
    this.identifier, 
    this.datetimeConverter,
    this.addressHasher,
    this.createAddress,
    this.addressType,
    this.networkTimestampClass
  );
  
  Uint8List publicKeyToAddress(PublicKey publicKey) {
    var partOneHashBuilder = addressHasher();
    var partOneHash = partOneHashBuilder.process(publicKey.bytes);
    var partTwoHash = RIPEMD160().update(Uint8List.fromList(partOneHash)).digest();
    var version = Uint8List.fromList([identifier, ...partTwoHash]);
    var partThreeHashBuilder = addressHasher();
    var checksum = Uint8List.fromList(partThreeHashBuilder.process(version).sublist(0, 4));
    return createAddress(version, checksum);
  }
}

abstract class AddressConstructable {
  AddressConstructable();
  static const int ENCODED_SIZE = 0;
}

abstract class Constructable {
  Constructable();
}

File: sdk/php/src/Network/Network.php
<?php

namespace SymbolSdk\Network;
use ReflectionClass;

class Network
{
  public string $name;
  public $identifier;
  public NetworkTimestampDatetimeConverter $datetimeConverter;
  protected $_addressHasher;
  protected $_createAddress;
  protected ReflectionClass $_addressClass;
  public ReflectionClass $networkTimestampClass;
  public function __construct($name, $identifier, $datetimeConverter, $addressHasher, $createAddress, $addressClass, $networkTimestampClass)
  {
    $this->name = $name;
    $this->identifier = $identifier;
    $this->datetimeConverter = $datetimeConverter;
    $this->_addressHasher  = $addressHasher;
    $this->_createAddress  = $createAddress;
    $this->_addressClass = new ReflectionClass($addressClass);
    $this->networkTimestampClass = new ReflectionClass($networkTimestampClass);
  }

  public function publicKeyToAddress(string $publicKey)
  {
    $partOneHash = hash($this->_addressHasher, $publicKey, true);
    $partTwoHash = hash('ripemd160', $partOneHash, true);
    $identifierBinary = pack('C', $this->identifier);
    $version = $identifierBinary . $partTwoHash;
    $partThreeHash = hash($this->_addressHasher, $version, true);
    $checksum = substr($partThreeHash, 0, 4);
    return call_user_func($this->_createAddress, $version, $checksum);
  }
}

